<a href="https://colab.research.google.com/github/ksetdekov/ml_dpo_2021/blob/master/hw/hw3/hw03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание №3

Задание выполнил(а): _Сетдеков Кирилл_

### Общая информация

__Дата выдачи:__ 29.05.2021

__Дедлайн:__ 07.06.2021

### О задании

Часть 1 содержит теоретические задачи на линейные методы классификации и метрики качества классификации.

Часть 2 содержит практические задания на применение линейных методов классификации, подсчет метрик качества классификации и анализ результатов.

Обратите внимание, что вам не только нужно написать код, но и в некоторых местах ответить на вопросы.

### Оценивание и штрафы


#### Theory  [10 баллов]
* [Задание 1](#task1) [1 балл]
* [Задание 2](#task2) [1 балл]
* [Задание 3](#task3) [1 балл]
* [Задание 4](#task4) [1 балл]
* [Задание 5](#task5) [1 балл]
* [Задание 6](#task6) [1 балл]
* [Задание 7](#task7) [1 балл]
* [Задание 8](#task8) [1 балл]
* [Задание 9](#task9) [1 балл]
* [Задание 10](#task10) [1 балл]

#### Linear Classification [13 баллов]
* [Задание 1](#task2_1) [1 балл]
* [Задание 2](#task2_2) [1 балл]
* [Задание 3](#task2_3) [1 балл]
* [Задание 4](#task2_4) [1 балл]
* [Задание 5](#task2_5) [1 балл]
* [Задание 6](#task2_6) [1 балл]
* [Задание 7](#task2_7) [1 балл]
* [Задание 8](#task2_8) [1 балл]
* [Задание 9](#task2_9) [1 балл]
* [Задание 10](#task2_10) [1 балл]
* [Задание 11](#task2_11) [1 балл]
* [Задание 12](#task2_12) [1 балл]
* [Задание 13](#task2_13) [1 балл]


Итоговая оценка за домашнюю работу вычисляется по формуле: $$s \cdot \frac{10}{23},$$ где $s$ - сумма набранных балов. 

За сдачу задания позже срока на итоговую оценку за задание накладывается штраф в размере 0.25 балла в день (от оценки в 10 бальной шкале), но получить отрицательную оценку нельзя.

__Внимание!__ Домашнее задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов.

### Формат сдачи

Загрузка файлов с решениями происходит в системе [Anytask](https://anytask.org/course/811). 

# Часть 1. Теоретическая

In [26]:
# import all
import numpy as np
import pandas as pd
import multiprocessing

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from sklearn import svm

import matplotlib.pyplot as plt

### Задание 1 <a id="task1"></a>

Какой из представленных графиков соответствует графику функции потерь в логистической регрессии? Тот же вопрос про метод опорных векторов.

![loss.png](https://github.com/ksetdekov/ml_dpo_2021/raw/master/hw/hw3/loss.png)

**Ответ:**
- Логистическая регрессия: график № 2, оранжевый
- Метод опорных векторов: график № 3, зеленый

### Задание 2 <a id="task2"></a>

При решении задачи бинарной классификации алгоритм на отложенной выборке верно предсказывает 100 из 400 объектов класса 0 и 75 из 100 объектов класса 1. Найдите долю правильных ответов данного классификатора.

In [2]:
175/500

0.35

**Ответ:** Всего есть $Total= N+P=400+100=500$ объектов, из которых верно предсказаны $TN+TP = 100+75=175$ штук. Вопрос доли правильных ответов классификатора - найти accuracy.
$$ACC=\frac{TN+TP}{Total}=\frac{175}{500}=0.35$$

### Задание 3 <a id="task3"></a>

Точность алгоритма классификации равна 0.6, а полнота равна 0.9. Чему в таком случае равно значение F-меры?

In [3]:
2*(0.6*0.9)/(0.6+0.9)

0.7200000000000001

Нам даны точность (precision) и полнота (recall).
$$F=2\frac{precision\cdot recall}{precision+recall}=2\frac{0.6\cdot0.9}{0.6+0.9}=0.72$$

**Ответ:** 0.72


### Задание 4 <a id="task4"></a>

При решении задачи бинарной классификации использовались данные с двумя признаками. На данных была обучена логистическая регрессия. Известно, что после обучения вектор весов $(w_0, w_1, w_2) = (0.2, 0.7, 0.1)$ (вес $w_0$ соответствует единичному признаку). Для объекта $x$ с признаковым описанием $(3, 5)$ посчитайте модуль отступа.

_Не перепутайте с отступом классификатора - шириной разделяющей гиперплоскости. Это термин, более применимый к SVM. А отступ объекта (то, что требуется тут) - это немного другое, он показывает, ошибается ли классификатор на данном объекте и насколько._

In [4]:
weights_4 = [0.2, 0.7, 0.1]
x_4 = [1, 3, 5]
otstup = np.dot(weights_4, x_4)
print(otstup)

2.8


Отступ - это скалярное произведение весов на признаки, умноженное на индикатор класса. Модуль отступа будет положительный и будет равен скалярному произведению $<w,x_i> = <(0.2,0.7,0.1),(1,3,5)> = 2.8$

**Ответ:** 2.8


### Задание 5 <a id="task5"></a>

При решении задачи бинарной классификации использовались данные с двумя признаками. На данных была обучена логистическая регрессия. Известно, что после обучения вектор весов $(w_0, w_1, w_2) = (0.2, 0.7, 0.1)$ (вес $w_0$ соответствует единичному признаку). Для объекта $x$ с признаковым описанием $(3, 5)$ вычислите предсказание вероятности принадлежности классу 1. Ответ округлите до двух знаков после запятой.

In [5]:
print(f'p = {1 / (1+np.exp(-otstup)):.2f}')

p = 0.94


 $$P(y_i = 1) = \sigma(<w,x_i>)=\frac{1}{1+exp(-<w,x_i>)}=\frac{1}{1+exp(-
 2.8)}=0.94$$

 **Ответ:**0.94

### Задание 6 <a id="task6"></a>

Какой из представленных графиков соответствует обученной логистической регрессии, для которой не настраивался свободный коэффициент - то есть $w_0 = 0$ (в `sklearn` это гиперпараметр `fit_intercept = False`)?

![logregs.png](https://raw.githubusercontent.com/ksetdekov/ml_dpo_2021/master/hw/hw3/img6.png)

**Ответ:** график № 2, так как разделяющая прямая проходит через точку 0, 0 для двух признаков.

### Задание 7 <a id="task7"></a>

В задаче имеется 100 объектов. У каждого объекта 20 признаков. Сколько параметров (весов) будет настроено логистической регрессией в процессе обучения, учитывая свободный коэффициент?

**Ответ:** Будет построено 21 параметра (20 признаков и 1 свободный член)

### Задание 8 <a id="task8"></a>

Вы решаете задачу бинарной линейной классификации в трехмерном признаковом пространстве. Соответственно, решающее правило имеет вид:

$$
a(x) = \text{sign}(w_0 + w_1x_1 + w_2x_2 + w_3x_3)
$$

У вас есть следующая выборка:

| $x_1$ | $x_2$ | $x_3$ | $y$ |
|------|------|------|-----|
| 0.2 | 0.4 | 0 | +1|
| 0.5 | 0.9 | 0 | +1|
| 0.3 | 0.3 | 0 | +1|
| 0.1 | 0.8 | 1 | -1|
| 0.5 | 0.7 | 1 | -1|
| 0.9 | 0.9 | 1 | -1|
| 0.1 | 0.3 | 1 | -1|

а) Сколько существует наборов коэффициентов $w = (w_0, w_1, w_2, w_3)$, при которых задача будет решена идеально (с accuracy = 1)?

б) Если ответ в п. а) - 0, пропустите это задание. Если ответ в п. а) - 1 или 2, запишите этот набор коэффициентов (или два набора соответственно). Иначе приведите хотя бы три таких набора коэффициентов.

In [6]:
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score
import pandas as pd
Xss = [[0.2, 0.4, 0],
       [0.5, 0.9, 0],
       [0.3, 0.3, 0],
       [0.1, 0.8, 1],
       [0.5, 0.7, 1],
       [0.9, 0.9, 1],
       [0.1, 0.3, 1]]
X = pd.DataFrame(Xss ,columns = ['x1','x2','x3'])
y = np.array([1,1,1,-1,-1,-1,-1])


clf = SGDClassifier(loss='log', learning_rate='constant', eta0=0.1, random_state=13)
clf.fit(X, y)
y_pred_sgdclf = clf.predict(X)

print(f'accuracy is {accuracy_score(y, y_pred_sgdclf)}')
clf.coef_, clf.intercept_


accuracy is 1.0


(array([[ 0.23572938,  0.22551457, -4.96556729]]), array([1.89441577]))

In [7]:
ones = np.ones((7, 1))
with_ones = np.append(ones, Xss, axis=1)

weights_8high = [ 2, 0.23572938,  0.22551457, -4.96556729]
print(f'higher w0 predictions = {np.sign(np.dot(with_ones, weights_8high))}')
weights_8low = [ 1.8, 0.23572938,  0.22551457, -4.96556729]
print(f'lower w0 predictions = {np.sign(np.dot(with_ones, weights_8low))}')
weights_8fit = np.concatenate((clf.intercept_, clf.coef_[0]))
print(f'sgdfit = {np.sign(np.dot(with_ones, weights_8fit))}')

higher w0 predictions = [ 1.  1.  1. -1. -1. -1. -1.]
lower w0 predictions = [ 1.  1.  1. -1. -1. -1. -1.]
sgdfit = [ 1.  1.  1. -1. -1. -1. -1.]


**Ответ:**

а) существует бесконечно много наборов коэффициентов, которымми можно получить 100% accuracy для этих данных так как классы линейно разделимы множеством способов.

б) примером трех наборов $w = (w_0, w_1, w_2, w_3)$ будут:
  * оценка весов, найденная SGDClassifier : $(1.89441577, 0.23572938,  0.22551457, -4.96556729)$
  * оценка с более низким значением параметра  $w_0: (1.8, 0.23572938,  0.22551457, -4.96556729)$
  * оценка с более высоким значением параметра  $w_0: (2, 0.23572938,  0.22551457, -4.96556729)$

  все 3 дают точность 100%, выше построены прогнозы классов, которые совпадают с исходными.

### Задание 9 <a id="task9"></a>

Петя на работе построил алгоритм бинарной классификации $a(x)$, AUC-ROC которого равен 0.96. После этого Петя пошел обедать, а его коллега Вася, решив пошутить над Петей, сел за его компьютер и изменил код алгоритма так, чтобы вместо положительного класса алгоритм всегда предсказывал отрицательный, и наоборот – вместо отрицательного положительный. Какое значение AUC-ROC алгоритма обнаружил Вася, вернувшись с обеда?

Так как идеальный классификатор даст AUC 1, а идеально угадывающий наборот - 0, то происходит вычетание из 1 результата, если мы меняем классы с отрицательного на положительный. В этом примере: $1-0.96=0.04$

**Ответ:** 0.04

### Задание 10 <a id="task10"></a>

Александр построил алгоритм бинарной классификации $a(x)$, AUC-ROC которого равен 0.66. Михаил построил алгоритм $b(x) = \frac{a(x)}{3}$. То есть для каждого объекта алгоритм $b(x)$ выдает вероятность, которая в три раза меньше, чем вероятность, выданная алгоритмом $a(x)$. Чему равно значение AUC-ROC алгоритма $b(x)$?

**Ответ:** Так как сам алгоритм не поменялся, значение AUC-ROC не изменилось и будет равно 0.66. Будет отличаться в 3 раза меньшую сторону порог, при котором TPR(b) = TPR(a)

# Часть 2. Практическая

В этом задании вы будете работать с [данными о пациентах, у части которых есть заболевание сердца](https://www.kaggle.com/ronitf/heart-disease-uci). Вам нужно построить модель, классифицирующую пациентов на больных этим заболеванием и тех, у кого его нет, а также проанализировать результаты.

In [8]:
data = pd.read_csv('https://raw.githubusercontent.com/ksetdekov/ml_dpo_2021/master/hw/hw3/heart.csv')
data.head()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

Пациенты описываются следующими признаками:

- age
- sex
- chest pain type (4 values)
- resting blood pressure
- serum cholestoral in mg/dl
- fasting blood sugar > 120 mg/dl
- resting electrocardiographic results (values 0,1,2)
- maximum heart rate achieved
- exercise induced angina
- oldpeak = ST depression induced by exercise relative to rest
- the slope of the peak exercise ST segment
- number of major vessels (0-3) colored by flourosopy
- thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

Целевая переменная:

- target (0 or 1)

### Задание 1 <a id="task2_1"></a>

Какой процент пациентов из представленных в данных имеет заболевание сердца (`'target' = 1`)?

In [9]:
data['target'].mean()



0.5445544554455446

-----

Разделим данную выборку случайным образом на обучающую и тестовую части в отношении 3:1.

In [10]:
X_train, X_test, y_train, y_test = train_test_split(data.drop('target', axis=1), data['target'], test_size=0.25, random_state=13)
X_train.shape, X_test.shape

((227, 13), (76, 13))

### Задание 2 <a id="task2_2"></a>

Обучите логистическую регрессию из `sklearn` (`sklearn.linear_model.LogisticRegression`) со следующими гиперпараметрами:

   - максимальное число итераций: `max_iter=1000`
   - без регуляризации: `penalty='none'`
   - сид: `random_state=13`
   - остальные гиперпараметры оставьте равными по умолчанию

Подробнее о параметрах можно почитать в [документации](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).

Какое значение свободного коэффициента у полученной логистической регрессии? А какое получилось значение остальных весов?

In [11]:
logreg = LogisticRegression(penalty='none', random_state=13, max_iter=1000)

logreg.fit(X_train, y_train)
print(f'intercept = {logreg.intercept_[0]}')

valzip = zip(X_train.columns, logreg.coef_[0]) 
named_coef_base = pd.DataFrame(valzip, columns=['feature', 'base_log_coef'])
named_coef_base

intercept = 4.1642918393778166


feature  base_log_coef
0        age      -0.033380
1        sex      -1.905554
2         cp       0.745921
3   trestbps      -0.008394
4       chol      -0.003511
5        fbs       0.269271
6    restecg       0.491175
7    thalach       0.017390
8      exang      -1.104463
9    oldpeak      -0.459451
10     slope       0.827951
11        ca      -0.764477
12      thal      -1.024302

### Задание 3 <a id="task2_3"></a>

Какое значение L2-нормы вектора весов (без учета свободного коэффициента) у полученного линейного классификатора?

_Напоминание. L2-норма вектора $v = (v_1, \ldots, v_n)$ - это корень из суммы квадратов его элементов:_

$$
\|v\|_2 = \sqrt{\sum\limits_{i=1}^nv_i^2}
$$

In [12]:
print(f'L2-норма вектора весов: {np.linalg.norm(logreg.coef_):.4f}')

L2-норма вектора весов: 2.8728


### Задание 4 <a id="task2_4"></a>

Найдите долю правильных ответов (accuracy) классификатора на тестовой части выборки.

In [13]:
y_hat_test = logreg.predict(X_test)

print(f'доля правильных ответов на тесте: {accuracy_score(y_test, y_hat_test):.4f}')

доля правильных ответов на тесте: 0.8158


### Задание 5 <a id="task2_5"></a>

В задаче классификации, как и в задаче регрессии, для оптимизации линейных моделей можно применять регуляризацию. Этот метод реализован и в `sklearn.linear_model.LogisticRegression` - гиперпараметр, отвечающий за регуляризацию - это `C`, принимающее положительные значения. Класс реализован таким образом, что `C` является обратным значением к коэффициенту регуляризации - то есть чем меньше `C`, тем сильнее регуляризация, и наоборот. За тип регуляризации (нет регуляризации, L1, L2 или обе сразу) отвечает параметр `penalty`. Также для справки - есть гиперпараметр `solver`, задающий алгоритм, решающий оптимизационную проблему. Такие алгоритмы умеют работать не со всеми типами регуляризации, так что если, например, вы хотите обучить логистическую регрессию с L1-регуляризацией, но оставите гиперпараметр `solver` равным по умолчанию (`'lbfgs'`), то у вас возникнет ошибка (можете проверить).

Обучите логистическую регрессию заново со следующими гиперпараметрами:
   
   - максимальное число итераций: `max_iter=1000`
   - с L1-регуляризацией: `penalty='l1'`
   - с коэффициентом `C=0.1`
   - алгоритм оптимизации: `solver='liblinear'`
   - сид: `random_state=13`
   - остальные гиперпараметры оставьте равными по умолчанию
   
Отличается ли качество полученного классификатора от качества первого? Какая доля правильных ответов получается теперь на тестовой выборке?

In [14]:
logreg_l1 = LogisticRegression(penalty='l1', C=0.1, random_state=13, solver='liblinear', max_iter=1000)
logreg_l1.fit(X_train, y_train)

y_hatl1_test = logreg_l1.predict(X_test)

print(f'доля правильных ответов на тесте для L1 регуляризации: {accuracy_score(y_test, y_hatl1_test):.4f}')

print(f'ROC AUC исходная модель: {roc_auc_score(y_test, logreg.predict_proba(X_test)[:, 1]):.4f}')

print(f'ROC AUC для L1 регуляризации: {roc_auc_score(y_test, logreg_l1.predict_proba(X_test)[:, 1]):.4f}')



доля правильных ответов на тесте для L1 регуляризации: 0.8289
ROC AUC исходная модель: 0.8943
ROC AUC для L1 регуляризации: 0.9020


Доля правильных ответов на тесте выроста с 0,816 до 0,829.
По метрике ROC AUC, также происходит улучшение модели на тесте с 0.89 до 0,90

### Задание 6 <a id="task2_6"></a>

Выведите значения весов новой обученной логистической регрессии и сравните их с весами изначальной логистической регрессии. Что вы наблюдаете?

Найдите L2-норму вектора весов (без учета свободного коэффициента) для полученного классификатора. Как на нее повлияла регуляризация?

In [15]:
print(f'intercept = {logreg_l1.intercept_[0]}')
valzipl1 = zip(X_train.columns, logreg_l1.coef_[0]) 
named_coef_base
named_coef_l1 = pd.DataFrame(valzipl1, columns=['feature', 'L2_coef'])

result = pd.merge(named_coef_base, named_coef_l1)
print(f'L2-норма вектора весов: {np.linalg.norm(logreg_l1.coef_):.4f}')
result

intercept = 0.0
L2-норма вектора весов: 0.9158


feature  base_log_coef   L2_coef
0        age      -0.033380 -0.011195
1        sex      -1.905554 -0.246296
2         cp       0.745921  0.540721
3   trestbps      -0.008394 -0.008950
4       chol      -0.003511 -0.002121
5        fbs       0.269271  0.000000
6    restecg       0.491175  0.000000
7    thalach       0.017390  0.025371
8      exang      -1.104463  0.000000
9    oldpeak      -0.459451 -0.382568
10     slope       0.827951  0.000000
11        ca      -0.764477 -0.402679
12      thal      -1.024302 -0.419917

Содержательно - с L1 регуляризацией, пересечение и 4 коэффициента стали нулевыми. L2-норма вектора весов стала 0.9158 (против 2.87) на исходной оценке. $\Rightarrow$ регуляризация работает так как мы ожидаем, минимизуруя веса.

### Задание 7 <a id="task2_7"></a>

Построим третью модель - метод опорных векторов (SVM). В `sklearn` этот метод реализован в классе [`sklearn.svm.SVC`](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html). Мы проходили только линейную версию SVM (где мы рисуем гиперплоскость, разделяющую два класса) - ей соответствует гиперпараметр `kernel='linear'`.

Обучите SVM со следующими гиперпараметрами:

   - с коэффициентом `C=1`
   - с линейным ядром: `kernel='linear'`
   - сид: `random_state=13`
   - остальные гиперпараметры оставьте равными по умолчанию

Посчитайте долю правильных ответов построенного классификатора на тестовой выборке. Отличается ли оно от результатов двух построенных логистических регрессий?

In [16]:
svm_base = svm.SVC(C=1, kernel='linear', random_state=13)
svm_base.fit(X_train, y_train)

y_hatsvm_test = svm_base.predict(X_test)

print(f'доля правильных ответов на тесте для базового SVM: {accuracy_score(y_test, y_hatsvm_test):.4f}')

доля правильных ответов на тесте для базового SVM: 0.8421


Мы получили долю правильных ответов, которая выше (0,8421), чем у двух вариантов логистических регрессий (от 0,816 до 0,829).

### Задание 8 <a id="task2_8"></a>

Наконец, проверьте, как полученные классификаторы предсказывают не классы, а вероятности классов. В принципе, не все классификаторы позволяют предсказывать вероятности классов, но логистическая регрессия умеет это делать. Посмотрите на вероятности, которые выдает первая логистическая регрессия (обученный без регуляризации) на тестовой части выборки. В этом вам поможет метод [`predict_proba`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.predict_proba). Результатом его работы будет список размера $\ell\times 2$, где $\ell$ - это число объектов. В каждом столбце списка находятся вероятности соответствующего класса для объектов. Поэтому если вам нужен положительный класс, вас интересует последний столбец.

Какое получается значение AUC-ROC?

In [17]:
print(f'ROC AUC исходной логистической регрессии : {roc_auc_score(y_test, logreg.predict_proba(X_test)[:, 1]):.4f}')


ROC AUC исходной логистической регрессии : 0.8943


### Задание 9 <a id="task2_9"></a>

Посчитайте значение AUC-ROC для второй логистической регрессии (которая была обучена с L1-регуляризацией). Отличается ли оно от AUC-ROC первой логистической регрессии?

In [18]:

print(f'ROC AUC для L1 регуляризации: {roc_auc_score(y_test, logreg_l1.predict_proba(X_test)[:, 1]):.4f}')

ROC AUC для L1 регуляризации: 0.9020


Да, регуляризация улучшает качество модели, ROC AUC увеличивается.

### Задание 10 <a id="task2_10"></a>

Наконец, посчитаем значение AUC-ROC для метода опорных векторов. По умолчанию SVM не умеет выдавать вероятности. Однако вспомните, что значение AUC-ROC зависит только от того, как отранжированы объекты - но не от самих значений, выдаваемых классификатором. Поэтому на самом деле AUC-ROC необязательно подсчитывать только для вероятностей - его можно посчитать, даже если классификатор выдает вещественные числа не из интервала $[0, 1]$.

В случае SVM мы можем посчитать расстояние от объектов до разделяющей гиперплоскости (со знаком минус, если объект относится к отрицательному классу, и плюс - если к положительному). В `sklearn` это реализовано с помощью метода [`decision_function`](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html#sklearn.svm.SVC.decision_function). В отличие от метода `predict_proba` в логистической регрессии, данный метод выдает не массив размера $\ell\times 2$, а массив из $\ell$ элементов.

Примените метод `decision_function` для уже обученного SVM на тестовой выборке. Посмотрите на то, что выдает этот метод.

Посчитайте значение AUC-ROC для полученного результата. Что в итоге в этой задаче сработало лучше - SVM или логистическая регрессия?

In [19]:
print(f'ROC AUC для базовой SVM модели: {roc_auc_score(y_test, svm_base.decision_function(X_test)):.4f}')


ROC AUC для базовой SVM модели: 0.9035


На данном примере, на тестовой выборке, подход SVM показал более высокую долю верных ответов и более высокое значение AUC ROC чем два варианта логистических регрессий

### Задание 11 <a id="task2_11"></a>

Подберите оптимальные параметры модели SVM на кросс-валидации (используя разные значения параметра `C`). Для сравнения разных моделей используйте класс `GridSearchCV`, кросс-валидацию с 5 фолдами и ROC-AUC в качестве критерия качества. Выведите значени `C` лучшей модели.

In [31]:
svm_grid_searcher = GridSearchCV(
    svm.SVC(random_state=13),
    param_grid={
        'C': np.logspace(-1, 2, num=100),
        'kernel':['rbf', 'linear']
    },
    scoring='roc_auc',
    n_jobs=multiprocessing.cpu_count(),
    cv=5,
    verbose=10
)

svm_grid_searcher.fit(X_train, y_train)

Fitting 5 folds for each of 200 candidates, totalling 1000 fits


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done   1 tasks      | elapsed:    1.2s
[Parallel(n_jobs=2)]: Done   4 tasks      | elapsed:    1.3s
[Parallel(n_jobs=2)]: Done   9 tasks      | elapsed:    1.5s
[Parallel(n_jobs=2)]: Done  14 tasks      | elapsed:    1.5s
[Parallel(n_jobs=2)]: Batch computation too fast (0.1730s.) Setting batch_size=2.
[Parallel(n_jobs=2)]: Batch computation too fast (0.0947s.) Setting batch_size=4.
[Parallel(n_jobs=2)]: Done  24 tasks      | elapsed:    1.8s
[Parallel(n_jobs=2)]: Batch computation too fast (0.1652s.) Setting batch_size=8.
[Parallel(n_jobs=2)]: Done  58 tasks      | elapsed:    2.7s
[Parallel(n_jobs=2)]: Done 130 tasks      | elapsed:    5.5s
[Parallel(n_jobs=2)]: Done 202 tasks      | elapsed:    9.3s
[Parallel(n_jobs=2)]: Batch computation too slow (2.0983s.) Setting batch_size=1.
[Parallel(n_jobs=2)]: Done 269 tasks      | elapsed:   16.1s
[Parallel(n_jobs=2)]: Done 280 tasks      | elap

GridSearchCV(cv=5, error_score=nan,
             estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                           class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='scale', kernel='rbf', max_iter=-1,
                           probability=False, random_state=13, shrinking=True,
                           tol=0.001, verbose=False),
             iid='deprecated', n_jobs=2,
             param_grid={'C': array([  0.1       ,   0.10722672,   0.1149757 ,   0.12328467...
        26.56087783,  28.48035868,  30.53855509,  32.74549163,
        35.11191734,  37.64935807,  40.37017259,  43.28761281,
        46.41588834,  49.77023564,  53.36699231,  57.22367659,
        61.35907273,  65.79332247,  70.54802311,  75.64633276,
        81.11308308,  86.97490026,  93.26033469, 100.        ]),
                         'kernel': ['rbf', 'linear']},
             pre_dispatch='2*n_jobs', refit=True, return_trai

Значение C и kernel для лучшей модели:

In [32]:
svm_grid_searcher.best_params_

{'C': 49.770235643321136, 'kernel': 'linear'}

### Задание 12 <a id="task2_12"></a>

Для лучшей модели из предыдущего задания посчитайте ROC-AUC на тестовой выборке.

In [33]:
print(f'ROC AUC для оптимизированной SVM модели: {roc_auc_score(y_test, svm_grid_searcher.decision_function(X_test)):.4f}')


ROC AUC для оптимизированной SVM модели: 0.8978


В автоматическом режиме через gridsearchcv получился результат чуть хуже, чем для случайно выбранного параметра C=1, потенциально мы считаем что модель полученная будет лучше генерализироваться.

### Задание 13 <a id="task2_13"></a>

Для лучшей модели из предыдущего задания сопоставьте признаки в данных со значениями весов обученной модели (метод `.coef_`). 

Посчитайте, какой признак является самым важным для принятия решения (то есть имеет наибольший по модулю коэффициент). Выведите название соответствующего столбца.

In [34]:
valzipsvm = zip(X_train.columns, svm_grid_searcher.best_estimator_.coef_[0]) 

named_coef_svm = pd.DataFrame(valzipsvm, columns=['feature', 'svm_coef'])
named_coef_svm.sort_values(by=['svm_coef'], ascending=False, inplace=True, key=lambda x:  np.absolute(x))
named_coef_svm

feature  svm_coef
1        sex -2.194166
8      exang -1.698202
10     slope  1.177178
12      thal -1.163242
6    restecg  0.996712
11        ca -0.940348
2         cp  0.915909
9    oldpeak -0.487694
5        fbs -0.105566
0        age -0.024497
7    thalach  0.011726
3   trestbps -0.011625
4       chol -0.007184

In [35]:
print(f'Самый важный фактор для модели получился: {named_coef_svm.feature[0]}')

Самый важный фактор для модели получился: age


In [36]:
from sklearn.metrics import classification_report, confusion_matrix  
print(classification_report(y_test,svm_grid_searcher.predict(X_test)))#Output
print(confusion_matrix(y_test,svm_grid_searcher.predict(X_test)))


              precision    recall  f1-score   support

           0       0.81      0.79      0.80        33
           1       0.84      0.86      0.85        43

    accuracy                           0.83        76
   macro avg       0.83      0.82      0.83        76
weighted avg       0.83      0.83      0.83        76

[[26  7]
 [ 6 37]]
